cpp/src/arrow/dataset/filter.h

```diff
@@ -305,6 +310,37 @@ class ARROW_DS_EXPORT NotExpression final
   Result<std::shared_ptr<DataType>> Validate(const Schema& schema) const override;
 };
 
+class ARROW_DS_EXPORT IsValidExpression final
+    : public ExpressionImpl<UnaryExpression, IsValidExpression,
+                            ExpressionType::IS_VALID> {
+ public:
+  using ExpressionImpl::ExpressionImpl;
+
+  std::string ToString() const override;
+
+  Result<std::shared_ptr<DataType>> Validate(const Schema& schema) const override;
+
+  std::shared_ptr<Expression> Assume(const Expression& given) const override;
+};
+
+class ARROW_DS_EXPORT InExpression final
+    : public ExpressionImpl<UnaryExpression, InExpression, ExpressionType::IN> {
+ public:
+  InExpression(std::shared_ptr<Array> set, std::shared_ptr<Expression> operand)
```

Comment:
Not important, but if it's "operand IN set", then perhaps `(operand, set)` would be a more natural argument order.

Reply:
will do